Initial pass at an `init` command

stevegeek opened the pull request:
As discussed in #144 an init command might be nice. 

This is probably the first time I have ever written any Rust so I just tried copying things I saw in the existing implementation and did some work arounds to compile errors using ChatGPT. 

Ideally someone with some experience could help turn this into a proper solution that is more idiomatic etc!

I added the config file creation at a point before the configuration load happens to avoid it blowing up but Im sure there is a better solution here. 

Anyway thought it might be a starting point!

alexevanczuk reviewed (commented):
This looks good! Want to add an integration test?
You can check out some of the examples for inspiration.

stevegeek commented:
Will do!

stevegeek commented:
@alexevanczuk apologies for the delay on this, I will pick it up again asap

alexevanczuk commented:
No rush @stevegeek ! Totally on your schedule here :)

stevegeek commented:
@alexevanczuk made some changes to the way I had implemented this command. Think this implementation solves a few of the issues with the previous one. Added a test too

alexevanczuk commented on src/packs.rs:
```diff
@@ -49,6 +49,65 @@ pub fn greet() {
+pub fn init(absolute_root: &PathBuf) -> anyhow::Result<()> {
+    let root_package = "\
+# This file represents the root package of the application
+# Please validate the configuration using `packwerk validate` (for Rails applications) or running the auto generated
```
I think this is fine for the first iteration 👍🏼 I think for folks setting up `pks` and plan to use it entirely instead of `packwerk`, it might be interesting in the future to replace `packwerk` with `pks`. We'd probably need another flag for that though since other commands rely on the existence of `packs.yml` to determine if we're in `packs_first_mode`, but neither file would be there at the time this command is used.

alexevanczuk commented on src/packs.rs:
```diff
@@ -49,6 +49,65 @@ pub fn greet() {
+# cache_directory: \"tmp/cache/packwerk\"
+";
+    let root_package_path = absolute_root.join("package.yml");
+    let packs_config_path = absolute_root.join("packwerk.yml"); // Can you just create a packs.yml?
```
I think we should actually only write one for now... I'm thinking just writing `packwerk.yml` makes sense, in the interest of this gem being a strict port of `packwerk` to start?

On the other hand, I can easily see the argument that anyone using `init` with `pks` maybe should just begin in `packs_first_mode` from the beginning, what do you think?

stevegeek replied:
I think it prob makes sense for `pks init` to generate `packs.yml`, unless a switch says othewise as you mentioned in your other comment. Would be happy to drop a CLI switch in if I can work it out

alexevanczuk commented on src/packs/cli.rs:
```diff
@@ -186,6 +189,13 @@ pub fn run() -> anyhow::Result<()> {
 
     install_logger(args.debug);
 
+    match args.command {
```
Any reason not to have `packs::init` be called in the case statement below?

stevegeek replied:
This was because `configuration = packs::configuration::get(&absolute_root)` blows up if you set the root path to and configuration-less directory (as you would in the case of `init`ing in a new project) - `No root pack found. First double check a root pack exists`. I was not sure how to restructure everything to allow for this fact (ie that before an `init` command there isnt a configuration) ... so I put the processing of the command before the configuration load... hacky but not well versed enough in Rust to think of a better solution!

alexevanczuk reviewed (commented):
Looks good, one thought that we should just create one YML file to start. I kind of like the idea of just creating the `packs.yml` and referencing `pks` commands instead of packwerk commands, since if someone is starting a new app for the first time it's a good time to start using `pks` from the beginning, what do you think?

Also you have a couple of lint errors but if you run `cargo fmt && cargo clippy` it should show you where they are! There are some cmd line flags to auto fix too (you can have this set up in your editor to auto-fix lint errors).

stevegeek commented:
> Looks good, one thought that we should just create one YML file to start. I kind of like the idea of just creating the `packs.yml` and referencing `pks` commands instead of packwerk commands, since if someone is starting a new app for the first time it's a good time to start using `pks` from the beginning, what do you think?

Makes sense to assume one would be using `pks` from the get go in this case. I will look at packs first mode, didnt even notice it before, have always used a `packwerk.yml`

 
> Also you have a couple of lint errors but if you run `cargo fmt && cargo clippy` it should show you where they are! There are some cmd line flags to auto fix too (you can have this set up in your editor to auto-fix lint errors).

Ok, yeah I saw there was a githook but it was blowing up and I couldn't work out how to get it to run so committed without it. will make sure to run the formatter next time!

stevegeek commented on src/packs/cli.rs:
```diff
@@ -186,6 +193,10 @@ pub fn run() -> anyhow::Result<()> {
 
     install_logger(args.debug);
 
+    if let Command::Init { use_packwerk } = args.command {
```
Same reason as mentioned above -> https://github.com/alexevanczuk/packs/pull/149#discussion_r1721817511 , not sure how I would restructure everything given that this command will run without any configuration available

alexevanczuk commented:
Nice!! Just two very small changes requested:
1. Can ya add that inline comment about why the case on the command is higher for init?
2. Want to bump the patch version in `Cargo.toml`? That will allow the new version to be auto-published when merged 🎉

alexevanczuk reviewed (approved):
Nice, thank you for the contribution and congrats on your first Rust PR!!
